### PR TITLE
Email templates API v2 implementation

### PR DESCRIPTION
**Pull request by Avarjana**
### Description
Improve email templates API to store application wise branding preferences for email templates.

### Related issue
- https://github.com/wso2/product-is/issues/19853

### Todo
- [x] bump `identity-event-handler-notification` version from https://github.com/wso2-extensions/identity-event-handler-notification/pull/223

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml [hunk 1]
```diff
@@ -0,0 +1,809 @@
...
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '403':
+          $ref: '#/components/responses/Forbidden'
+        '409':
+          $ref: '#/components/responses/Conflict'
+        '500':
+          $ref: '#/components/responses/ServerError'
+  '/email/template-types/{template-type-id}':
+    get:
```
[the remark this answers is not in the document]
Comment (Avarjana) [reply components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml:108]: Addressed with https://github.com/wso2/identity-api-server/pull/595/commits/472164a2d6bfcee39e2879ca6d54a4cc86c590e3

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml [hunk 2]
```diff
@@ -0,0 +1,809 @@
...
+          type: string
+          example: WSO2 Identity Server Team
+          description: The footer of the email.
+    EmailTemplateWithID:
+      allOf:
+        - $ref: '#/components/schemas/EmailTemplate'
+        - required:
+            - id
+          properties:
+            id:
```
[the remark this answers is not in the document]
Comment (Avarjana) [reply components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml]: Addressed with https://github.com/wso2/identity-api-server/pull/595/commits/472164a2d6bfcee39e2879ca6d54a4cc86c590e3

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml [hunk 3]
```diff
@@ -0,0 +1,799 @@
...
+          $ref: '#/components/responses/InvalidInput'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '403':
+          $ref: '#/components/responses/Forbidden'
+        '404':
+          $ref: '#/components/responses/NotFound'
+        '500':
+          $ref: '#/components/responses/ServerError'
+  '/email/template-types/{template-type-id}/org-templates':
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml:167]: Check if we can move this to a bulk endpoint or remove this. Same with the app-templates path.

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/java/org/wso2/carbon/identity/rest/api/server/email/template/v2/impl/EmailApiServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    @Autowired
+    private ServerEmailTemplatesService emailTemplatesService;
+
+    @Autowired
+    private ApplicationEmailTemplatesService applicationEmailTemplatesService;
+
+    @Override
+    public Response addAppEmailTemplate(
+            String templateTypeId, String appUuid, EmailTemplateWithID emailTemplateWithID) {
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/java/org/wso2/carbon/identity/rest/api/server/email/template/v2/impl/EmailApiServiceImpl.java:54]: Check if the params are sanitized and secure before moving to service layers

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml [hunk 5]
```diff
@@ -0,0 +1,799 @@
...
+        - Email Template Types
+      summary: Retrieves all the email template types.
+      operationId: getAllEmailTemplateTypes
+      description: |
+        Retrieves all the email template types in the system. <br>
+
+        <b>Permission required:</b><br>
+        * /permission/admin/manage/identity/emailmgt/view
+        <br> 
+        <b>Scopes required:</b>
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml:39]: Check if we need new scopes for applications

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml [hunk 6]
```diff
@@ -0,0 +1,799 @@
...
+          description: The body of the email.
+        footer:
+          type: string
+          example: WSO2 Identity Server Team
+          description: The footer of the email.
+    EmailTemplateWithID:
+      allOf:
+        - $ref: '#/components/schemas/EmailTemplate'
+        - required:
+            - locale
```
[the remark this answers is not in the document]
Comment (Avarjana) [reply components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml:705]: Let's focus on this when we move to the relational DB model.

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml (hunk 3 above)
Comment (Avarjana) [reply components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml:167]: Removed from https://github.com/wso2/identity-api-server/pull/595/commits/5ccba2c271c0d5ccfed72be993661c0c364df066

File: components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml (hunk 5 above)
Comment (Avarjana) [reply components/org.wso2.carbon.identity.api.server.email.template/org.wso2.carbon.identity.rest.api.server.email.template.v2/src/main/resources/email-template.yml:39]: Not addressing this ATM. If required, can be configured from the pack itself.

**Review by AnjanaSamindraPerera — APPROVED** [review]

**Pull request merged**